internal/operations/manager.go

```diff
@@ -0,0 +1,145 @@
+		return nil, i18n.NewError(ctx, i18n.MsgInitializationNilDepError)
+	}
+	om := &operationsManager{
+		ctx:      ctx,
+		database: di,
+		tokens:   ti,
+		handlers: make(map[fftypes.OpType]OperationHandler),
+	}
+	return om, nil
+}
+
+func (om *operationsManager) RegisterHandler(handler OperationHandler, ops []fftypes.OpType) {
+	for _, opType := range ops {
+		om.handlers[opType] = handler
+	}
+}
+
+func (om *operationsManager) PrepareOperation(ctx context.Context, op *fftypes.Operation) (*fftypes.PreparedOperation, error) {
+	handler, ok := om.handlers[op.Type]
+	if !ok {
+		return nil, i18n.NewError(ctx, i18n.MsgOperationNotSupported)
+	}
+	return handler.PrepareOperation(ctx, op)
+}
+
+func (om *operationsManager) RunOperation(ctx context.Context, op *fftypes.PreparedOperation) error {
+	handler, ok := om.handlers[op.Type]
+	if !ok {
+		return i18n.NewError(ctx, i18n.MsgOperationNotSupported)
+	}
+	if complete, err := handler.RunOperation(ctx, op); err != nil {
+		om.writeOperationFailure(ctx, op.ID, err)
+		return err
+	} else if complete {
+		om.writeOperationSuccess(ctx, op.ID)
+	}
+	return nil
+}
+
+func (om *operationsManager) findLatestRetry(ctx context.Context, opID *fftypes.UUID) (op *fftypes.Operation, err error) {
+	op, err = om.database.GetOperationByID(ctx, opID)
+	if err != nil {
+		return nil, err
+	}
+	if op.Retry == nil {
+		return op, nil
+	}
+	return om.findLatestRetry(ctx, op.Retry)
+}
+
+func (om *operationsManager) RetryOperation(ctx context.Context, ns string, opID *fftypes.UUID) (op *fftypes.Operation, err error) {
+	var po *fftypes.PreparedOperation
+	err = om.database.RunAsGroup(ctx, func(ctx context.Context) error {
+		op, err = om.findLatestRetry(ctx, opID)
+		if err != nil {
+			return err
+		}
+
+		// Create a copy of the operation with a new ID
+		op.ID = fftypes.NewUUID()
+		op.Status = fftypes.OpStatusPending
+		op.Error = ""
+		op.Output = nil
+		op.Created = fftypes.Now()
+		op.Updated = op.Created
+		if err = om.database.InsertOperation(ctx, op); err != nil {
+			return err
+		}
+
+		// Update the old operation to point to the new one
+		update := database.OperationQueryFactory.NewUpdate(ctx).Set("retry", op.ID)
+		if err = om.database.UpdateOperation(ctx, opID, update); err != nil {
+			return err
+		}
+
+		po, err = om.PrepareOperation(ctx, op)
+		return err
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return op, om.RunOperation(ctx, po)
+}
+
+func (om *operationsManager) writeOperationSuccess(ctx context.Context, opID *fftypes.UUID) {
+	if err := om.database.ResolveOperation(ctx, opID, fftypes.OpStatusSucceeded, "", nil); err != nil {
+		log.L(ctx).Errorf("Failed to update operation %s: %s", opID, err)
+	}
+}
+
+func (om *operationsManager) writeOperationFailure(ctx context.Context, opID *fftypes.UUID, err error) {
+	if err := om.database.ResolveOperation(ctx, opID, fftypes.OpStatusFailed, err.Error(), nil); err != nil {
+		log.L(ctx).Errorf("Failed to update operation %s: %s", opID, err)
```

Comment:
As above: Think it would be good to write the full data of the operation here - particularly the outputs - as they would have been lost.

Reply:
Same as above - the only "outputs" from a synchronous failure are in the form of a Go `error` which is logged here. I'm not sure there's anything else useful to log... unless it's a sign that we're not returning enough info from some other layer.